iOS 18 does not allow a client application to specify SameSite=None

charliewolfe opened the pull request:
#### 45ce92c14115b2f8d56bc87301c6f1097a6f06cf
<pre>
iOS 18 does not allow a client application to specify SameSite=None
<a href="https://bugs.webkit.org/show_bug.cgi?id=279153">https://bugs.webkit.org/show_bug.cgi?id=279153</a>
<a href="https://rdar.apple.com/135312438">rdar://135312438</a>

Reviewed by Pascoe and Alex Christensen.

In the iOS 18 and macOS Sequoia betas, CFNetwork began treating cookies as SameSite=Lax by default. This
caused an issue where, when WebKit converts from its internal enum `Cookie::SameSitePolicy` to
`NSHTTPCookieStringPolicy`, we return nil for `SameSitePolicy::None`, which CFNetwork now interprets as
lax instead of none.

WebKit should explicitly return none when converting `WebCore::Cookie` to an `NSHTTPCookie`. We should
use a string constant for none, but one is not currently available in the SDK.

* Source/WTF/wtf/PlatformHave.h:
* Source/WebCore/platform/network/cocoa/CookieCocoa.mm:
(WebCore::coreSameSitePolicy):
(WebCore::nsSameSitePolicy):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKHTTPCookieStore.mm:
(TEST(WKHTTPCookieStore, SetSameSiteCookiePolicies)):
(TEST(WKHTTPCookieStore, SetSameSiteCookiePoliciesNonPersistentStore)):

Canonical link: <a href="https://commits.webkit.org/283230@main">https://commits.webkit.org/283230@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/232cadddfca49fe108865e46ffe66d9fcd9f407f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/65614 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/44991 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/18232 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/69640 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/16223 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/67732 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/52786 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/16510 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/52686 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/11266 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/68681 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/41549 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/56794 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/33313 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/38228 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/14173 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/15099 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/58728 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/60064 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/14514 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/71346 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/64858 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/9579 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/13960 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/60003 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/9600 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/56859 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/60277 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/7907 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/1556 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/86625 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/9945 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/40795 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/15249 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/41871 "Built successfully") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/43054 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/41615 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebCore/platform/network/cocoa/CookieCocoa.mm:
```diff
@@ -97,7 +97,7 @@ static NSHTTPCookieStringPolicy _Nullable nsSameSitePolicy(Cookie::SameSitePolic
     switch (policy) {
     case Cookie::SameSitePolicy::None:
-        return nil;
+        return @"none"; // FIXME: This should use a string constant (rdar://135227277)
```
Should coreSameSitePolicy also check for "none"?

charliewolfe replied:
None will be used if there is no match, but I should check “none” so we don’t hit the debug assertion in that function.

achristensen07 commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKHTTPCookieStore.mm:
```diff
@@ -817,6 +817,30 @@ static bool areCookiesEqual(NSHTTPCookie *first, NSHTTPCookie *second)
+            finished = true;
+        }];
+    }];
+    TestWebKitAPI::Util::run(&finished);
```
Could you call clearCookies again after this test to leave no state behind?
Also, is it worth adding the same test with a non-persistent data store?
Also, could we test the actual functionality of a same-site:none cookie and see that it is sent correctly over the network?

charliewolfe replied:
> Could you call clearCookies again after this test to leave no state behind?

Yes.

> Also, is it worth adding the same test with a non-persistent data store?

Sure.

> Also, could we test the actual functionality of a same-site:none cookie and see that it is sent correctly over the network?

Can you clarify what you mean by this? Do you want me to do that in this test? There are several existing layout tests that verify SameSite=None cookies are sent correctly over the network.

achristensen07 commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKHTTPCookieStore.mm:
```diff
@@ -817,6 +817,47 @@ static bool areCookiesEqual(NSHTTPCookie *first, NSHTTPCookie *second)
     TestWebKitAPI::Util::run(&done);
 }
 
+#if __IPHONE_OS_VERSION_MIN_REQUIRED >= 180000 || __MAC_OS_X_VERSION_MIN_REQUIRED >= 150000
```
What about tvOS, watchOS, visionOS?  This should probably be in PlatformHave.h instead.

achristensen07 reviewed (approved):
This test is sufficient.

charliewolfe committed:
iOS 18 does not allow a client application to specify SameSite=None
https://bugs.webkit.org/show_bug.cgi?id=279153
rdar://135312438

Reviewed by Pascoe and Alex Christensen.

In the iOS 18 and macOS Sequoia betas, CFNetwork began treating cookies as SameSite=Lax by default. This
caused an issue where, when WebKit converts from its internal enum `Cookie::SameSitePolicy` to
`NSHTTPCookieStringPolicy`, we return nil for `SameSitePolicy::None`, which CFNetwork now interprets as
lax instead of none.

WebKit should explicitly return none when converting `WebCore::Cookie` to an `NSHTTPCookie`. We should
use a string constant for none, but one is not currently available in the SDK.

* Source/WTF/wtf/PlatformHave.h:
* Source/WebCore/platform/network/cocoa/CookieCocoa.mm:
(WebCore::coreSameSitePolicy):
(WebCore::nsSameSitePolicy):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKHTTPCookieStore.mm:
(TEST(WKHTTPCookieStore, SetSameSiteCookiePolicies)):
(TEST(WKHTTPCookieStore, SetSameSiteCookiePoliciesNonPersistentStore)):

Canonical link: https://commits.webkit.org/283230@main
